File: examples/example_9.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

#include "agdeblend.h"

int main(void) {
  size_t n_traces[7];
  size_t n_times = 512;
  int n_patches = 7;
  int volumes[7] = {0};
  int n_dims[1] = {3};
  int **window_shapes = (int **)malloc(n_patches * sizeof(int *));
  int **coords = (int **)malloc(n_patches * sizeof(int *));
  int **shapes = (int **)malloc(n_patches * sizeof(int *));
  long int **shottimes = (long int **)malloc(n_patches * sizeof(long int *));
  int **channels = (int **)malloc(n_patches * sizeof(int *));
  enum AGDTraceType **trace_types =
      (enum AGDTraceType **)malloc(n_patches * sizeof(enum AGDTraceType *));
  float **data = (float **)malloc(n_patches * sizeof(float *));
  float initial_factor = 1.0f;
  int n_its = 1000;
  int print_freq = -1;
  int patch_idx;
  char filename[256];
  FILE *fid;
  int space_shape_all[2] = {32, 40}; /* 40 = 2 * 24 - 8 */
  int n_traces_all = space_shape_all[0] * space_shape_all[1];
  int file_range[2][2][2] = {{{0, 16}, {0, 24}}, {{16, 32}, {16, 40}}};
  int file_shape[2] = {16, 24};
  int n_traces_file = file_shape[0] * file_shape[1];
  int patch_ranges_x[3][2] = {{0, 16}, {8, 24}, {16, 32}};
  int patch_ranges_y[3][2] = {{0, 16}, {8, 32}, {24, 40}};
  int **patch_ranges = malloc(2 * n_patches * sizeof(int *));
  int ix;
  int iy;
  int trace_idx;
  int file_idx;
  long int *shottimes_file =
      (long int *)malloc(n_traces_file * sizeof(long int));
  int *channels_file = (int *)malloc(n_traces_file * sizeof(int));
  enum AGDTraceType *trace_types_file =
      (enum AGDTraceType *)malloc(n_traces_file * sizeof(enum AGDTraceType));
  float *data_file = (float *)malloc(n_traces_file * n_times * sizeof(float));
  long int *shottimes_all = (long int *)malloc(n_traces_all * sizeof(long int));
  int *channels_all = (int *)malloc(n_traces_all * sizeof(int));
  enum AGDTraceType *trace_types_all =
      (enum AGDTraceType *)malloc(n_traces_all * sizeof(enum AGDTraceType));
  float *data_all = (float *)malloc(n_traces_all * n_times * sizeof(float));

  for (trace_idx = 0; trace_idx < n_traces_all; trace_idx++) {
    trace_types_all[trace_idx] = AGDMissing;
  }

  window_shapes[0] = (int *)malloc(n_dims[0] * sizeof(int));
  window_shapes[0][0] = 16;
  window_shapes[0][1] = 16;
  window_shapes[0][2] = 256;

  patch_idx = 0;
  for (ix = 0; ix < 3; ix++) {
    for (iy = 0; iy < 3; iy++) {
      if ((ix == 2 && iy == 0) || (ix == 0 && iy == 2)) continue;
      patch_ranges[2 * patch_idx] = patch_ranges_x[ix];
      patch_ranges[2 * patch_idx + 1] = patch_ranges_y[iy];
      coords[patch_idx] = (int *)malloc((n_dims[0] - 1) * sizeof(int));
      coords[patch_idx][0] = ix;
      coords[patch_idx][1] = iy;
      shapes[patch_idx] = (int *)malloc(n_dims[0] * sizeof(int));
      shapes[patch_idx][0] =
          patch_ranges[2 * patch_idx][1] - patch_ranges[2 * patch_idx][0];
      shapes[patch_idx][1] = patch_ranges[2 * patch_idx + 1][1] -
                             patch_ranges[2 * patch_idx + 1][0];
      shapes[patch_idx][2] = n_times;
      n_traces[patch_idx] = shapes[patch_idx][0] * shapes[patch_idx][1];
      patch_idx++;
    }
  }

  for (patch_idx = 0; patch_idx < n_patches; patch_idx++) {
    shottimes[patch_idx] =
        (long int *)malloc(n_traces[patch_idx] * sizeof(long int));
    channels[patch_idx] = (int *)malloc(n_traces[patch_idx] * sizeof(int));
    trace_types[patch_idx] = (enum AGDTraceType *)malloc(
        n_traces[patch_idx] * sizeof(enum AGDTraceType));
    data[patch_idx] =
        (float *)malloc(n_traces[patch_idx] * n_times * sizeof(float));
  }

  /* Load from input files and add to _all arrays */
  for (file_idx = 0; file_idx < 2; file_idx++) {
    int file_start_x = file_range[file_idx][0][0];
    int file_start_y = file_range[file_idx][1][0];

    sprintf(filename, "data_4_shottimes_%d.bin", file_idx);
    fid = fopen(filename, "rb");
    fread(shottimes_file, sizeof(long int), n_traces_file, fid);
    fclose(fid);

    sprintf(filename, "data_4_channels_%d.bin", file_idx);
    fid = fopen(filename, "rb");
    fread(channels_file, sizeof(int), n_traces_file, fid);
    fclose(fid);

    sprintf(filename, "data_4_trace_types_%d.bin", file_idx);
    fid = fopen(filename, "rb");
    fread(trace_types_file, sizeof(enum AGDTraceType), n_traces_file, fid);
    fclose(fid);

    sprintf(filename, "data_4_blended_data_%d.bin", file_idx);
    fid = fopen(filename, "rb");
    fread(data_file, sizeof(float), n_traces_file * n_times, fid);
    fclose(fid);

    for (ix = 0; ix < file_shape[0]; ix++) {
      int trace_idx_all =
          (file_start_x + ix) * space_shape_all[1] + file_start_y;
      int trace_idx_file = ix * file_shape[1];
      memcpy(shottimes_all + trace_idx_all, shottimes_file + trace_idx_file,
             file_shape[1] * sizeof(long int));
      memcpy(channels_all + trace_idx_all, channels_file + trace_idx_file,
             file_shape[1] * sizeof(int));
      memcpy(trace_types_all + trace_idx_all, trace_types_file + trace_idx_file,
             file_shape[1] * sizeof(enum AGDTraceType));
      memcpy(data_all + trace_idx_all * n_times,
             data_file + trace_idx_file * n_times,
             file_shape[1] * n_times * sizeof(float));
    }
  }

  /* Extract patches from the _all arrays */
  for (patch_idx = 0; patch_idx < n_patches; patch_idx++) {
    int patch_start_x = patch_ranges[2 * patch_idx][0];
    int patch_start_y = patch_ranges[2 * patch_idx + 1][0];
    for (ix = 0; ix < shapes[patch_idx][0]; ix++) {
      int trace_idx_all =
          (patch_start_x + ix) * space_shape_all[1] + patch_start_y;
      int trace_idx_patch = ix * shapes[patch_idx][1];
      memcpy(shottimes[patch_idx] + trace_idx_patch,
             shottimes_all + trace_idx_all,
             shapes[patch_idx][1] * sizeof(long int));
      memcpy(channels[patch_idx] + trace_idx_patch,
             channels_all + trace_idx_all, shapes[patch_idx][1] * sizeof(int));
      memcpy(trace_types[patch_idx] + trace_idx_patch,
             trace_types_all + trace_idx_all,
             shapes[patch_idx][1] * sizeof(enum AGDTraceType));
      memcpy(data[patch_idx] + trace_idx_patch * n_times,
             data_all + trace_idx_all * n_times,
             shapes[patch_idx][1] * n_times * sizeof(float));
    }
  }

  /* Free the _file and _all arrays */
  free(shottimes_file);
  free(channels_file);
  free(trace_types_file);
  free(data_file);
  free(shottimes_all);
  free(channels_all);
  free(trace_types_all);
  free(data_all);

  /* Deblend */
  agd_deblend(n_patches, volumes, n_dims, (int const *const *)window_shapes,
              (int const *const *)coords, (int const *const *)shapes,
              (long int const *const *)shottimes, (int const *const *)channels,
              (enum AGDTraceType const *const *)trace_types, NULL, NULL, NULL,
              initial_factor, n_its, print_freq, data);

  /* Allocate data_all and add patches to it */
  data_all = (float *)calloc(n_traces_all * n_times, sizeof(float));
  for (patch_idx = 0; patch_idx < n_patches; patch_idx++) {
    int patch_start_x = patch_ranges[2 * patch_idx][0];
    int patch_start_y = patch_ranges[2 * patch_idx + 1][0];
    for (ix = 0; ix < shapes[patch_idx][0]; ix++) {
      for (iy = 0; iy < shapes[patch_idx][1]; iy++) {
        int trace_idx_all =
            (patch_start_x + ix) * space_shape_all[1] * n_times +
            (patch_start_y + iy) * n_times;
        int trace_idx_patch =
            ix * shapes[patch_idx][1] * n_times + iy * n_times;
        int time_idx;
        for (time_idx = 0; time_idx < n_times; time_idx++) {
          data_all[trace_idx_all + time_idx] +=
              data[patch_idx][trace_idx_patch + time_idx];
        }
      }
    }
  }

  /* Allocate data_file, extract data for each file and write out */
  data_file = (float *)malloc(n_traces_file * n_times * sizeof(float));
  for (file_idx = 0; file_idx < 2; file_idx++) {
    int file_start_x = file_range[file_idx][0][0];
    int file_start_y = file_range[file_idx][1][0];
    for (ix = 0; ix < file_shape[0]; ix++) {
      int trace_idx_all = (file_start_x + ix) * space_shape_all[1] * n_times +
                          file_start_y * n_times;
      int trace_idx_file = ix * file_shape[1] * n_times;
      memcpy(data_file + trace_idx_file, data_all + trace_idx_all,
             file_shape[1] * n_times * sizeof(float));
    }
    sprintf(filename, "out/data_4_deblended_data_%d.bin", file_idx);
    fid = fopen(filename, "wb");
    fwrite(data_file, sizeof(float), n_traces_file * n_times, fid);
    fclose(fid);
  }
  free(data_file);
  free(data_all);

  free(window_shapes[0]);
  for (patch_idx = 0; patch_idx < n_patches; patch_idx++) {
    free(coords[patch_idx]);
    free(shapes[patch_idx]);
    free(shottimes[patch_idx]);
    free(channels[patch_idx]);
    free(trace_types[patch_idx]);
    free(data[patch_idx]);
  }
  free(window_shapes);
  free(coords);
  free(shapes);
  free(shottimes);
  free(channels);
  free(trace_types);
  free(data);

  return 0;
}

File: examples/example_9.f90
program example_9

use, intrinsic :: iso_c_binding, only: c_int, c_long, c_float
use :: agdeblend_m, only: deblend, deblend_patch, volume, AGDMissing

implicit none

type(deblend_patch), dimension(:), allocatable :: patches
type(volume), dimension(:), allocatable :: volumes
integer, parameter :: n_times = 512
integer, parameter :: n_patches = 7
integer, parameter :: n_window_traces_x = 16
integer, parameter :: n_window_traces_y = 16
integer, parameter :: n_window_times = 256
real, parameter :: initial_factor = 1.0
integer, parameter :: n_its = 1000
integer, parameter :: print_freq = -1
integer :: fid
integer :: patch_idx
integer :: file_idx
integer :: ix
integer :: iy
character(len=256) :: filename
integer :: ierr
integer, parameter :: space_shape_all(2) = (/ 40, 32 /)
integer, parameter :: file_ranges(2, 2, 2) =                            &
  reshape(source = (/ 1, 24, 1, 16, 17, 40, 17, 32 /),                  &
          shape = shape(file_ranges))
integer(kind=c_long), allocatable :: shottimes_all(:, :)
integer(kind=c_int), allocatable :: channels_all(:, :)
integer(kind=c_int), allocatable :: trace_types_all(:, :)
real(kind=c_float), allocatable :: data_all(:, :, :)
integer, parameter :: patch_ranges_y(2, 3) =                            &
  reshape(source = (/ 1, 16, 9, 32, 25, 40 /),                          &
          shape = shape(patch_ranges_y))
integer, parameter :: patch_ranges_x(2, 3) =                            &
  reshape(source = (/ 1, 16, 9, 24, 17, 32 /),                          &
          shape = shape(patch_ranges_x))
integer :: patch_ranges(2, 2, 7)
integer :: n_traces_y
integer :: n_traces_x

allocate(shottimes_all(space_shape_all(1), space_shape_all(2)))
allocate(channels_all(space_shape_all(1), space_shape_all(2)))
allocate(trace_types_all(space_shape_all(1), space_shape_all(2)))
allocate(data_all(n_times, space_shape_all(1), space_shape_all(2)))

trace_types_all = AGDMissing

! Read input files
do file_idx = 1, 2
  write (filename, '(a, i0, a)') "data_4_shottimes_", file_idx - 1,     &
    ".bin"
  open(newunit=fid, file=trim(filename), form="unformatted",            &
       access="stream", action="read", status="old")
  read(fid) shottimes_all(file_ranges(1, 1, file_idx):                  &
                          file_ranges(2, 1, file_idx),                  &
                          file_ranges(1, 2, file_idx):                  &
                          file_ranges(2, 2, file_idx))
  close(fid)

  write (filename, '(a, i0, a)') "data_4_channels_", file_idx - 1,      &
    ".bin"
  open(newunit=fid, file=trim(filename), form="unformatted",            &
       access="stream", action="read", status="old")
  read(fid) channels_all(file_ranges(1, 1, file_idx):                   &
                         file_ranges(2, 1, file_idx),                   &
                         file_ranges(1, 2, file_idx):                   &
                         file_ranges(2, 2, file_idx))
  close(fid)

  write (filename, '(a, i0, a)') "data_4_trace_types_", file_idx - 1,   &
    ".bin"
  open(newunit=fid, file=trim(filename), form="unformatted",            &
       access="stream", action="read", status="old")
  read(fid) trace_types_all(file_ranges(1, 1, file_idx):                &
                            file_ranges(2, 1, file_idx),                &
                            file_ranges(1, 2, file_idx):                &
                            file_ranges(2, 2, file_idx))
  close(fid)

  write (filename, '(a, i0, a)') "data_4_blended_data_", file_idx - 1,  &
    ".bin"
  open(newunit=fid, file=trim(filename), form="unformatted",            &
       access="stream", action="read", status="old")
  read(fid) data_all(:,                                                 &
                     file_ranges(1, 1, file_idx):                       &
                     file_ranges(2, 1, file_idx),                       &
                     file_ranges(1, 2, file_idx):                       &
                     file_ranges(2, 2, file_idx))
  close(fid)
end do

! Set patches
allocate(patches(n_patches))
patch_idx = 1
do ix = 1, 3
  do iy = 1, 3
    if ((ix == 3 .and. iy == 1) .or. (ix == 1 .and. iy == 3)) then
      cycle
    end if
    patch_ranges(:, 1, patch_idx) = patch_ranges_y(:, iy)
    patch_ranges(:, 2, patch_idx) = patch_ranges_x(:, ix)
    n_traces_y = patch_ranges_y(2, iy) - patch_ranges_y(1, iy) + 1
    n_traces_x = patch_ranges_x(2, ix) - patch_ranges_x(1, ix) + 1
    patches(patch_idx)%volume_idx = 0
    allocate(patches(patch_idx)%coords(2))
    patches(patch_idx)%coords(1) = iy
    patches(patch_idx)%coords(2) = ix
    allocate(patches(patch_idx)%data_shape(3))
    patches(patch_idx)%data_shape(1) = n_times
    patches(patch_idx)%data_shape(2) = n_traces_y
    patches(patch_idx)%data_shape(3) = n_traces_x
    allocate(patches(patch_idx)%shottimes(n_traces_y * n_traces_x))
    allocate(patches(patch_idx)%channels(n_traces_y * n_traces_x))
    allocate(patches(patch_idx)%trace_types(n_traces_y * n_traces_x))
    allocate(patches(patch_idx)%values                                  &
      (n_times * n_traces_y * n_traces_x))
    patches(patch_idx)%shottimes(:) =                                   &
      reshape(source = shottimes_all(patch_ranges(1, 1, patch_idx) :    &
                                     patch_ranges(2, 1, patch_idx),     &
                                     patch_ranges(1, 2, patch_idx) :    &
                                     patch_ranges(2, 2, patch_idx)),    &
              shape = (/ n_traces_y * n_traces_x /))
    patches(patch_idx)%channels(:) =                                    &
      reshape(source = channels_all(patch_ranges(1, 1, patch_idx) :     &
                                    patch_ranges(2, 1, patch_idx),      &
                                    patch_ranges(1, 2, patch_idx) :     &
                                    patch_ranges(2, 2, patch_idx)),     &
              shape = (/ n_traces_y * n_traces_x /))
    patches(patch_idx)%trace_types(:) =                                 &
      reshape(source = trace_types_all(patch_ranges(1, 1, patch_idx) :  &
                                       patch_ranges(2, 1, patch_idx),   &
                                       patch_ranges(1, 2, patch_idx) :  &
                                       patch_ranges(2, 2, patch_idx)),  &
              shape = (/ n_traces_y * n_traces_x /))
    patches(patch_idx)%values(:) =                                      &
      reshape(source = data_all(:,                                      &
                                patch_ranges(1, 1, patch_idx) :         &
                                patch_ranges(2, 1, patch_idx),          &
                                patch_ranges(1, 2, patch_idx) :         &
                                patch_ranges(2, 2, patch_idx)),         &
              shape = (/ n_times * n_traces_y * n_traces_x /))
    patch_idx = patch_idx + 1
  end do
end do

deallocate(shottimes_all)
deallocate(channels_all)
deallocate(trace_types_all)
deallocate(data_all)

allocate(volumes(1))
volumes(1)%n_dims = 3
allocate(volumes(1)%window_shape(3))
volumes(1)%window_shape(1) = n_window_times
volumes(1)%window_shape(2) = n_window_traces_y
volumes(1)%window_shape(3) = n_window_traces_x

! Deblend
call deblend(patches, volumes, initial_factor, n_its, print_freq, ierr)

if (ierr /= 0) then
  stop 1
end if

! Sum patches
allocate(data_all(n_times, space_shape_all(1), space_shape_all(2)))
data_all = 0

do patch_idx = 1, n_patches
  n_traces_y = patch_ranges(2, 1, patch_idx) -                          &
               patch_ranges(1, 1, patch_idx) + 1
  n_traces_x = patch_ranges(2, 2, patch_idx) -                          &
               patch_ranges(1, 2, patch_idx) + 1
  data_all(:,                                                           &
           patch_ranges(1, 1, patch_idx):patch_ranges(2, 1, patch_idx), &
           patch_ranges(1, 2, patch_idx):patch_ranges(2, 2, patch_idx)) &
    = data_all(:,                                                       &
           patch_ranges(1, 1, patch_idx):patch_ranges(2, 1, patch_idx), &
           patch_ranges(1, 2, patch_idx):patch_ranges(2, 2, patch_idx)) &
    + reshape(source = patches(patch_idx)%values,                       &
              shape = (/ n_times, n_traces_y, n_traces_x /))
end do

! Extract file ranges and save to disk
do file_idx = 1, 2
  write (filename, '(a, i0, a)') "out/data_4_deblended_data_",          &
         file_idx - 1, ".bin"
  open(newunit=fid, file=trim(filename), form="unformatted",            &
       access="stream", action="write", status="replace")
  write(fid) data_all(:,                                                &
                      file_ranges(1, 1, file_idx):                      &
                      file_ranges(2, 1, file_idx),                      &
                      file_ranges(1, 2, file_idx):                      &
                      file_ranges(2, 2, file_idx))
  close(fid)
end do
deallocate(data_all)

do patch_idx = 1, n_patches
  deallocate(patches(patch_idx)%coords)
  deallocate(patches(patch_idx)%data_shape)
  deallocate(patches(patch_idx)%shottimes)
  deallocate(patches(patch_idx)%channels)
  deallocate(patches(patch_idx)%trace_types)
  deallocate(patches(patch_idx)%values)
end do
deallocate(patches)
deallocate(volumes(1)%window_shape)
deallocate(volumes)

end program example_9
